cfd_protocol/src/lib.rs

```diff
@@ -646,11 +648,110 @@ mod tests {
+                (taker_own_pk, taker_rev_pk, taker_publish_pk),
+            );
+
+            let commit_output = TxOut {
+                value: commit_amount.as_sat(),
+                script_pubkey: commit_descriptor
+                    .address(Network::Regtest)
+                    .expect("can derive address from descriptor")
+                    .script_pubkey(),
+            };
+
+            let inner = Transaction {
+                version: 2,
+                lock_time: 0,
+                input: vec![lock_input],
+                output: vec![commit_output],
+            };
+
+            Ok(Self {
+                inner,
+                commit_descriptor,
+                commit_amount,
+            })
+        }
+
+        fn commit_outpoint(&self) -> OutPoint {
+            let txid = self.inner.txid();
+            let vout = self
+                .inner
+                .output
+                .iter()
+                .position(|out| out.script_pubkey == self.commit_descriptor.script_pubkey())
+                .expect("to find dlc output in lock tx");
+
+            OutPoint {
+                txid,
+                vout: vout as u32,
+            }
+        }
+
+        fn build_commit_descriptor(
+            (maker_own_pk, maker_rev_pk, maker_publish_pk): (PublicKey, PublicKey, PublicKey),
+            (taker_own_pk, taker_rev_pk, taker_publish_pk): (PublicKey, PublicKey, PublicKey),
+        ) -> Descriptor<PublicKey> {
+            // TODO: Optimize miniscript
+
+            let maker_own_pk_hash = hash160::Hash::hash(&maker_own_pk.key.serialize()[..]);
+            let maker_own_pk = (&maker_own_pk.key.serialize().to_vec()).to_hex();
+            let taker_own_pk_hash = hash160::Hash::hash(&taker_own_pk.key.serialize()[..]);
+            let taker_own_pk = (&taker_own_pk.key.serialize().to_vec()).to_hex();
+
+            let maker_rev_pk_hash = hash160::Hash::hash(&maker_rev_pk.key.serialize()[..]);
+            let taker_rev_pk_hash = hash160::Hash::hash(&taker_rev_pk.key.serialize()[..]);
+
+            let maker_publish_pk_hash = hash160::Hash::hash(&maker_publish_pk.key.serialize()[..]);
+            let taker_publish_pk_hash = hash160::Hash::hash(&taker_publish_pk.key.serialize()[..]);
+
+            let cet_or_refund_condition =
+                format!("and_v(v:pk({}),pk_k({}))", maker_own_pk, taker_own_pk);
+            let maker_punish_condition = format!(
+                "and_v(v:pkh({}),and_v(v:pkh({}),pk_h({})))",
+                maker_own_pk_hash, taker_publish_pk_hash, taker_rev_pk_hash
+            );
+            let taker_punish_condition = format!(
+                "and_v(v:pkh({}),and_v(v:pkh({}),pk_h({})))",
+                taker_own_pk_hash, maker_publish_pk_hash, maker_rev_pk_hash
+            );
+            let descriptor_str = format!(
+                "wsh(c:or_i(or_i({},{}),{}))",
+                maker_punish_condition, taker_punish_condition, cet_or_refund_condition
+            );
```

Comment:
Have you considered using the miniscript compiler to create a miniscript from a policy instead of handrolling it? Might yield a more optimal version!

Reply:
I'm going to create an issue for this since the TODO disappeared :)

Reply:
I used the online compiler for the updated script.

Reply:
Ah, I didn't know that! It would still be good to have the code generate the miniscript from the policy, so I think we can leave the issue open.